### ISSUE TITLE
Gulp 4 - alpha2/alpha3 duplicate path bug

### ISSUE THREAD

**Issue by alohaninja** [gulpjs/gulp#2082]
**Bug Origin**
`gulp.src` is not removing duplicate paths with **gulp4 alpha3**, but in **gulp4 alpha2** it does. We concat ordered paths, then a glob path and it works as expected in **alpha2**. 

**What actually happened?**
Multiple paths are included when using `gulp.src`

**Please post a sample of your gulpfile (preferably reduced to just the bit that's not working)**

```
gulp.src(["./lib/angular.js", "./lib/**/*.js"]);
```

Mulitiple `/lib/angular.js` items are listed instead of it filtering duplicate paths. Verified the casing in the filesystem matches the casing in the `src`. 

**What version of gulp are you using?**
gulp4 alpha3

**What versions of npm and node are you using?**
npm 5.6 (also tried 5.2)

**Workaround**
Removing the local `node_modules` path under `node_modules\gulp\node_modules` temporarily fixes this issue, but when you run `npm i` again it re-builds this folder and creates the problem again. 
In alpha2, there was no `node_modules` subdirectory.

**Comment by phated** [comment]
Are those absolute paths? What operating system? What version of node? Can you provide more context than just a gulp.src call?

**Comment by alohaninja** [comment]
@phated using relative paths, Win10 OS, node 8.2.1. 

Use case for us is using `gulp-inject` for inserting static dependencies (css, js,etc.) into our asp.net layout page. This worked fine for gulp 3, we've been running gulp 4 alpha2 for awhile - someone just got a new machine and noticed that alpha3 broke it as alpha2 is no longer available.

**Comment by phated** [comment]
@alohaninja I think this has to be a problem on Windows because it's being excluded for me on OSX.  I'm going to need to do some weird appveyor stuff to try this on Windows.

**Comment by phated** [comment]
@alohaninja it seems to be working properly on AppVeyor (see https://ci.appveyor.com/project/phated/vfs-unique/build/3) - note that only 2 files are logged.

Can you run https://github.com/phated/vfs-unique on your machine?

**Comment by alohaninja** [comment]
@phated - thanks for setting up this simple example for isolating the issue. I was also unable to replicate it using your example also so I thought maybe it was something in our dependency chain. 

After further research it was discovered that our relative paths are sort of odd, but do contain multiple `./` within the path due to variable reuse within our gulp script. Modifying your example project to the following will reproduce the error that used to work pre-alpha3. 

### Culprit Path Issue (_multiple ./ within path_)
```
gulp.src(['./src/./file.js', './src/./**/*.js']);
```

I suspect this is not gulp-specific per-se, but maybe a lower-level dependency that has changed behavior.

**Comment by phated** [comment]
I'm pretty sure that's not a valid glob but I'll ask the authors of our globbing deps.

**Comment by phated** [comment]
@alohaninja I confirmed that is an invalid glob - however, we do seem to support it (due to one of our transitive dependencies) so I'm trying to find the proper fix. Thanks for bearing with me on this.

**Comment by jonschlinkert** [comment]
@alohaninja remove `./` from the globs. technically that's not supported and it's probably messing up your matching (minimatch has a couple of different issues open related to that).

> After further research it was discovered that our relative paths are sort of odd, but do contain multiple ./ within the path due to variable reuse within our gulp script.

To clarify, the path you mentioned (with `/./`) is created by your build script? Or is that path being created by gulp or a gulp dep?

**Comment by alohaninja** [comment]
@jonschlinkert @phated - agreed that this not a standard glob pattern, we just have been using this script for years now and not run into this until upgrading to gulp4 alpha3, others may run into it. 

The extra `/./` is created by our gulp script, it's not generated by gulp or gulp dep - it just used to work and now it doesn't. Definitely was hard to track down as we didn't change our script, we just updated the dependencies.

**Comment by phated** [comment]
@alohaninja to clarify your assumption: there should be no assumption made that upgrading to a new alpha or the published 4.0.0 should work - there have been breaking changes between each alpha and the published versions.  That being said, I want this behavior to be consistent between Windows and Posix, which it currently is not.

**Comment by phated** [comment]
From deeper digging, this is a non-trivial problem inside `node-glob` so I'm going to close this as such: __The behavior of invalid globs is undefined, especially cross-platform. Please fix your globs by removing non-standard patterns like `/./` or `/../`.__

@alohaninja You might also be interested in the `uniqueBy` option as supplied by glob-stream: https://github.com/gulpjs/glob-stream#optionsuniqueby
